### ISSUE TITLE
Differing dimensions for training and test

### ISSUE THREAD

**Issue by FelixNeutatz** [scikit-learn-contrib/categorical-encoding#91]
Hi,

I would like to fit encodings on my training set and then using this fitted encoding to transform both the training and the test set:

```
import category_encoders as ce

train = ['Brunswick East', 'Fitzroy', 'Williamstown', 'Newport', 'Balwyn North', 'Doncaster', 'Melbourne', 'Albert Park', 'Bentleigh', 'Northcote']
test = ['Fitzroy North', 'Fitzroy', 'Richmond', 'Surrey Hills', 'Blackburn', 'Port Melbourne', 'Footscray', 'Yarraville', 'Carnegie', 'Surrey Hills']

encoder = ce.HelmertEncoder()
encoder.fit(train)

train_t = encoder.transform(train)
test_t = encoder.transform(test)

print train_t.shape
>> (10, 10)
print test_t.shape
>> (10, 2)
```
The problem is that the dimensions do not fit. What do I do wrong or how can I fix this issue?

Best regards,
Felix

**Comment by janmotl** [comment]
Observations:
1) This happens not only when a testing set contains a _new_ value, but also when the testing set does not contain _every_ value from the training set.
2) The issue also affects: BackwardDifferenceEncoder, PolynomialEncoder and SumEncoder.

**Comment by FelixNeutatz** [comment]
So, is there some workaround to fix this issue?

**Comment by janmotl** [comment]
The only workaround I see is to encode training and testing data together. However, this does not solve the deployment of the encoder.

**Comment by FelixNeutatz** [comment]
Putting training and test data together would violate the principles of ML :(

**Comment by wdm0006** [comment]
You seem to be passing in just a single list? Pass in a pandas dataframe or a numpy array (2d array, so in this case one column) instead and see if you get the same effect. The encoders that pass thru to statsmodels haven't been tested with list in types.

**Comment by FelixNeutatz** [comment]
I get the same effect:
```
import category_encoders as ce
import numpy as np
import pandas as pd

train = ['Brunswick East', 'Fitzroy', 'Williamstown', 'Newport', 'Balwyn North', 'Doncaster', 'Melbourne', 'Albert Park', 'Bentleigh', 'Northcote']
test = ['Fitzroy North', 'Fitzroy', 'Richmond', 'Surrey Hills', 'Blackburn', 'Port Melbourne', 'Footscray', 'Yarraville', 'Carnegie', 'Surrey Hills']

train_pd = pd.DataFrame(np.matrix(train).T)
test_pd = pd.DataFrame(np.matrix(test).T)

encoder = ce.HelmertEncoder()
encoder.fit(train_pd)

train_t = encoder.transform(train_pd)
test_t = encoder.transform(test_pd)

print train_t.shape
>> (10, 10)
print test_t.shape
>> (10, 2)
```

**Comment by FelixNeutatz** [comment]
Does anybody else have a solution?

**Comment by JohnnyC08** [comment]
@janmotl I want to know if we should include the columns, but as zeroed out.

In the test case Felix provided only col_0_0 and col_0_1 are being returned in the test set. In that case, should the columns for 2-9 also be returned with zero value or is there some other default value that should be used. You'll have to forgive me I am not overly familiar with helmert encoding and have only just started learning it to jump in and figure this out.

If the columns should be zero I can submit a WIP PR that does this, but it is not the prettiest fix in the world, so we can work together to improve it. There might be an edge case I'm not sure about.

**Comment by janmotl** [comment]
All following encodings: BackwardDifference, Helmert, Polynomial and SumEncoder were designed that way that intercept has [zeroes](https://stats.idre.ucla.edu/spss/faq/coding-systems-for-categorical-variables-in-regression-analysis/) in all output columns.

Since it is reasonable to predict grand average for a categorical value that was never seen before, zeroes are reasonable.

**Comment by JohnnyC08** [comment]
@janmotl I guess I'm confused. From what you're telling me if there are unknown values passed into transform, then they should be involved in the calculation?

I've seen different encoders handle new cases differently. 
In the one hot encoder transform, new values are simply represented by a row of zeroes.
For the target encoder, unknown values are either represented as NaN, if impute missing is false, or the mean, if impute missing is true.

What I proposed for the helmert encoder is that missing values don't impact the calculation and we return columns of zeroes representing the values measured during the train component of the process. 

Below is one picture from the result of the train array Felix is mentioning.
<img width="911" alt="train_result" src="https://user-images.githubusercontent.com/10440346/46186761-1602d480-c295-11e8-8ebe-71644e7e1fe0.png">

For the test case, I have went ahead and made the below happen.
<img width="916" alt="test_result" src="https://user-images.githubusercontent.com/10440346/46186785-37fc5700-c295-11e8-87df-99af4626e56a.png">

Is this what we want to happen when unknown values are used in transform?

**Comment by janmotl** [comment]
Ah, I should have said "that a _downstream model_ should predict grand average when a new value is observed".  "Row of zeroes" fulfills this requirement.  

I wrote a few tests that should, in my opinion, pass:
```python
    def test_preserve_dimension(self):
        for encoder_name in encoders.__all__:
            with self.subTest(encoder_name=encoder_name):
                encoder = getattr(encoders, encoder_name)()
                train = encoder.fit_transform(X, y)
                test = encoder.transform(X_t)
                self.assertEqual(len(train.columns), len(test.columns), 'The count of attributes must not change')

    def test_helmert_preserve_dimension_1(self):
        train = ['A', 'B', 'C']
        test = ['A', 'D', 'E']

        encoder = encoders.HelmertEncoder()
        encoder.fit(train)
        test_t = encoder.transform(test)

        expected = [[1, -1, -1],
                    [1,  0,  0],
                    [1,  0,  0]]
        self.assertEqual(test_t.values.tolist(), expected)

    def test_helmert_preserve_dimension_2(self):
        train = ['A', 'B', 'C']
        test = ['B', 'D', 'E']

        encoder = encoders.HelmertEncoder()
        encoder.fit(train)
        test_t = encoder.transform(test)

        expected = [[1,  1, -1],
                    [1,  0,  0],
                    [1,  0,  0]]
        self.assertEqual(test_t.values.tolist(), expected)

    def test_helmert_preserve_dimension_3(self):
        train = ['A', 'B', 'C']
        test = ['A', 'B', 'C', None]

        encoder = encoders.HelmertEncoder()
        encoder.fit(train)
        test_t = encoder.transform(test)

        expected = [[1, -1, -1],
                    [1,  1, -1],
                    [1,  0,  2],
                    [1,  0,  0]]
        self.assertEqual(test_t.values.tolist(), expected)

    def test_helmert_preserve_dimension_4(self):
        train = ['A', 'B', 'C']
        test = ['D', 'B', 'C', None]

        encoder = encoders.HelmertEncoder()
        encoder.fit(train)
        test_t = encoder.transform(test)

        expected = [[1,  0,  0],
                    [1,  1, -1],
                    [1,  0,  2],
                    [1,  0,  0]]
        self.assertEqual(test_t.values.tolist(), expected)
```

It is likely that not all the tests are necessary and that some important tests missing... use your judgment. Furthermore, the tests represent my expectations, not the ground truth.

From quick testing:
1. Value "A" in the testing set should preferably be encoded just like in the training set -> it is not enough to just add constant zero column(s) into the encoded testing set.
2. It is important to preserve the "reference", to which the "contrasts" are calculated, between training and testing set.

A possible solution to the above issues: Store a dictionary with the encoding and use the dictionary in the `transform` step. Essentially, what LOO, TargetEncoder or WOE are doing. But it may not be the best solution.

**Comment by janmotl** [comment]
Solved with #128
